### PR TITLE
[logging, lora] feat: Extend count_flops to allow Lora config for Qwen 2/3/3.5 models

### PR DESCRIPTION
**Pull request by wyettzeng**
### What does this PR do?

Extend count_flops to allow Lora config for Qwen 2/3/3.5 models

### Checklist Before Starting

- Search for relative PRs/issues and link here: ...
- PR title follows `[{modules}] {type}: {description}` format (enforced by [check_pr_title.yml](.github/workflows/check_pr_title.yml))
  - **Allowed modules:** `agent`, `ci`, `ckpt`, `config`, `data`, `dist`, `docker`, `docs`, `logging`, `lora`, `misc`, `model`, `omni`, `optim`, `ops`, `parallel`, `perf`, `release`, `task`, `trainer`
  - **Allowed types:** `feat`, `fix`, `refactor`, `chore`, `test`
  - Breaking changes: prepend `[BREAKING]` — e.g. `[BREAKING][parallel, model] feat: dynamic batching`

### Test

FFT flops count unchanged for:
- 336 different Qwen models, each with 11 batch/sequence settings.
- Run a quick sanity check fo 0.6B qwen model + lora, and mfu is reported
- Added scripts/profile/compare_lora_flops.py, a script that uses Veomni's config loader to load config, and pass it to the flops counter. Tested 10 different Text / Image / MOE models, all output are similar to the following:
image

### API and Usage Example

> Show API changes and usage examples if applicable.

### Design & Code Changes

> High-level design description and specific change list.

### Checklist Before Submitting

- Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- Applied pre-commit checks
- Added/updated documentation
- If `tasks/` training scripts were moved or renamed: updated `docs/` examples and verified `python3 scripts/ci/check_doc_task_paths.py` passes (also enforced by the **Check doc task paths** CI workflow)
- Added tests to CI workflow (or explained why not feasible)

### CODE REVIEW

File: veomni/utils/count_flops.py [hunk 1]
```diff
@@ -22,6 +22,39 @@
...
+)
+
+_LORA_MODULES_BY_MODEL_TYPE = {
+    "qwen2": _QWEN2_LORA_MODULES,
+    "qwen3": _QWEN2_LORA_MODULES,
+    "qwen3_5": _QWEN3_5_LORA_MODULES,
+    "qwen3_5_text": _QWEN3_5_LORA_MODULES,
+}
+
+
```
Comment (xzzWZY) [inline veomni/utils/count_flops.py]: can we move these to `veomni/lora/config.py`？

Comment (wyettzeng) [reply veomni/utils/count_flops.py]: done

File: veomni/utils/count_flops.py [hunk 2]
```diff
@@ -961,8 +1061,66 @@ def _estimate_qwen3_5_family_flops(self, tokens_sum, batch_seqlens, delta_time,
...
+                # Qwen3.5-MoE currently has no fused-expert target converter.
+                # Unqualified MLP names therefore match the shared expert's
+                # nn.Linear layers, while routed expert parameters stay frozen.
+                shared_count = num_hidden_layers
+                module_shapes_and_counts.update(
+                    {
+                        "gate_proj": (hidden_size, text_config.shared_expert_intermediate_size, shared_count),
+                        "up_proj": (hidden_size, text_config.shared_expert_intermediate_size, shared_count),
+                        "down_proj": (text_config.shared_expert_intermediate_size, hidden_size, shared_count),
+                    }
```
Comment (xzzWZY) [inline veomni/utils/count_flops.py]: Do we support lora for experts which are not shared?

File: veomni/utils/count_flops.py [hunk 3]
```diff
@@ -884,8 +939,45 @@ def _estimate_qwen3_next_flops(self, tokens_sum, batch_seqlens, delta_time):
...
+                "out_proj": (linear_v_size, hidden_size, num_linear_attn_layers),
+                "gate_proj": (hidden_size, self.config.shared_expert_intermediate_size, shared_count),
+                "up_proj": (hidden_size, self.config.shared_expert_intermediate_size, shared_count),
+                "down_proj": (self.config.shared_expert_intermediate_size, hidden_size, shared_count),
+            }
+            lora_N = self._compute_lora_params(lora_rank, lora_modules, module_shapes_and_counts)
+            # The legacy helper omits q_proj's output-gate half. Include the
+            # real frozen projection shape when LoRA arithmetic is requested.
+            gated_q_proj_N = hidden_size * q_size * num_full_attn_layers
+            dense_N_flops = (4 * (base_N + gated_q_proj_N) + 6 * lora_N) * tokens_sum
```
Comment (xzzWZY) [inline veomni/utils/count_flops.py]: Why gated_q_proj_N not in base_N and it's not calculated in full finetuning?

File: veomni/utils/count_flops.py [hunk 4]
```diff
@@ -961,8 +1061,66 @@ def _estimate_qwen3_5_family_flops(self, tokens_sum, batch_seqlens, delta_time,
...
+                        "up_proj": (hidden_size, text_config.intermediate_size, num_hidden_layers),
+                        "down_proj": (text_config.intermediate_size, hidden_size, num_hidden_layers),
+                    }
+                )
+            lora_N = self._compute_lora_params(lora_rank, lora_modules, module_shapes_and_counts)
+
+            # The legacy helper ignores q_proj's output-gate half. Include it in
+            # LoRA mode so frozen-base and adapter work use the real projection.
+            gated_q_proj_N = hidden_size * q_size * num_full_attn_layers
+            base_N = mlp_N + attn_linear_N + gated_q_proj_N + lm_head_N
```
Comment (xzzWZY) [inline veomni/utils/count_flops.py]: same here, why "gated_q_proj_N" not included in full finetuning?

**Review by xzzWZY — APPROVED** [review]

File: veomni/utils/count_flops.py (hunk 4 above)
Comment (wyettzeng) [reply veomni/utils/count_flops.py]: fixed in another PR

File: veomni/utils/count_flops.py (hunk 3 above)
Comment (wyettzeng) [reply veomni/utils/count_flops.py]: fixed in another PR

File: veomni/utils/count_flops.py (hunk 2 above)
Comment (wyettzeng) [reply veomni/utils/count_flops.py]: just add in support for routed experts lora support

File: veomni/utils/count_flops.py [hunk 5]
```diff
@@ -961,8 +1061,66 @@ def _estimate_qwen3_5_family_flops(self, tokens_sum, batch_seqlens, delta_time,
...
+                    {
+                        "gate_proj": (hidden_size, text_config.intermediate_size, num_hidden_layers),
+                        "up_proj": (hidden_size, text_config.intermediate_size, num_hidden_layers),
+                        "down_proj": (text_config.intermediate_size, hidden_size, num_hidden_layers),
+                    }
+                )
+            lora_N = self._compute_lora_params(lora_rank, lora_modules, module_shapes_and_counts)
+
+            # The legacy helper ignores q_proj's output-gate half. Include it in
+            # LoRA mode so frozen-base and adapter work use the real projection.
```
[the remark this answers is not in the document]
Comment (wyettzeng) [reply veomni/utils/count_flops.py]: resolved

File: veomni/utils/count_flops.py [hunk 6]
```diff
@@ -15,6 +15,7 @@
 
 from transformers import PretrainedConfig
 
+from ..lora.config import LORA_MODULES_BY_MODEL_TYPE
```
[the remark this answers is not in the document]
Comment (wyettzeng) [reply veomni/utils/count_flops.py]: @xzzWZY asked me to put it there, so leaving there for now

File: veomni/utils/count_flops.py [hunk 7]
```diff
@@ -652,7 +679,7 @@ def _estimate_qwen3_vit_flop(self, images_seqlens, config):
         dense_N = patch_embed_N + (mlp_N + attn_linear_N) * depth + deepstack_merger_N + merger_N
 
         # non-attn all_layer & all_token fwd & bwd flops
-        dense_N_flops = 6 * dense_N * tokens_sum
+        dense_N_flops = (2 if frozen else 6) * dense_N * tokens_sum
```
[the remark this answers is not in the document]
Comment (wyettzeng) [reply veomni/utils/count_flops.py]: fixed, but it is worth noting that vlm_trainer currently does not support LoRA, so lora calculation here is more so added for future capabilities. Current logic:

input has not image token -> vit flops = 0
input has image token but none of the lora target modules point to vit -> forward only vit
otherwise -> vit lora training

another thing worth pointing out is that our lora currently only support Linear layers and moe experts component (expressed as nn.Embedding), so only Linear layer has lora flops counted. If we add in support for other modules like conv3d later, we will need to update count_flops.

**Comment by wyettzeng** [comment]
Reply to AI review:

1. Wire it into training (this is what makes the feature real) -> done
2. Don't raise from the metering path -> done, for unresolvable lora config we just report 0 flops as standard + log a warning for the user without crashing training
3. Fix the layering / single source of truth -> @xzzWZY asked to be put it there, but open to put it elsewhere if other reviewers would prefer
4. Model semantics that the current shape gets wrong — honour or explicitly document -> fixed, now take in VeOmniLoraConfig
5. missing docs -> fixed

File: tests/utils/test_count_flops.py
Comment (Luosuu) [inline tests/utils/test_count_flops.py:1]: do we have to add such a large test?

File: veomni/utils/count_flops.py [hunk 8]
```diff
@@ -104,7 +107,114 @@ def _compute_lm_head_params(hidden_size, vocab_size):
...
+            ]
+            if unsupported_patterns:
+                return (
+                    "LoRA FLOPs estimation only supports fused routed-expert target parameters ending in "
+                    f"{supported_suffixes}, got {unsupported_patterns!r}."
+                )
+        return None
+
+    @staticmethod
+    def _compute_routed_moe_lora_flops(config, tokens, lora_config):
```
Comment (Luosuu) [inline veomni/utils/count_flops.py:194]: I do not get what this is computing for

Comment (wyettzeng) [reply veomni/utils/count_flops.py:194]: For qwen 3.5 and 3 next, there are 2 types of experts:
- routed experts (ex: only topk expert active for each token), represented as nn.Embedding in code
- shared experts, where every token has to pass through it, represented as nn.Linear in code

qwen 3 only has routed experts. 

For veomni, routed expert can have 2 lora implementation: they all share the same AB lora matrix, or each get their own. This function calculates the flops specifically for these. If shared, the we can reuse the gate and up lora for all token, while need to have compute the down lora output for each expert independently as the input to down is no longer the same.

File: tests/utils/test_count_flops.py
Comment (wyettzeng) [reply tests/utils/test_count_flops.py:1]: reduced some, but I would recommend keeping the rest as this pr does add in quite a lot of new extra logic for lora, specifically moe, and vit have their own logic for different settings.

**Review by Luosuu — APPROVED** [review]
